src/clientdb/SnmpGadgets.h

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+/* DEBUG: section 00    Client Database */
+
+#ifndef _SQUID__SRC_CLIENTDB_SNMPGADGETS_H
+#define _SQUID__SRC_CLIENTDB_SNMPGADGETS_H
+
+#if SQUID_SNMP
+
+#include "ip/Address.h"
+#include "cache_snmp.h"
+#include "snmp_vars.h"
+
+/* AYJ 2021-12-20:
+ * take one step across the ClientDb cache map/hash from the given IP.
+ * If the cache entries have been changed between calls this will
+ * produce unexpected results;
+ * - all changes to the hash/map before 'current' will be skipped.
+ * - not finding 'current' will skip to the end and produce nullptr.
+ * XXX: use a better indexing mechanism for SNMP tables.
+ */
+const Ip::Address *client_entry(const Ip::Address *current);
```

Comment:
This PR already modifies all the callers so please split this function into `ClientDb::First(void)` and `ClientDb::Next(const Ip::Address &)`. This split will not fix all legacy problems with this API, of course, but it will noticeably reduce them by avoiding the (wrong) implication that the returned (and possibly dangling) pointer must be fed back to the function to continue iteration.